File: src/js/calculator.js
/**
 * Tactive Savings Calculator — Calculation Engine
 * Industry-average loss formulas and Tactive recovery rates
 */

// ── Industry Loss Rates ──
const LOSS_RATES = {
  materialWaste: 0.12,        // 12% of budget
  idleMachinery: 0.065,       // 6.5% of budget
  laborInefficiency: 0.20,    // 20% of labor cost
  scheduleOverrun: 0.15,      // 15% of budget
  reworkDefects: 0.07,        // 7% of budget
};

// ── Tactive Recovery Rates ──
const RECOVERY_RATES = {
  materialTracking: 0.70,     // Recovers 70% of material waste
  equipmentAlerts: 0.80,      // Recovers 80% of idle machinery cost
  laborDashboard: 0.60,       // Recovers 60% of labor inefficiency
  scheduleOptimizer: 0.50,    // Recovers 50% of schedule overrun
  qualityControl: 0.65,       // Recovers 65% of rework cost
};

// ── Constants ──
const AVG_DAILY_WAGE = 800;   // ₹800 per laborer per day (Indian avg)
const WORKING_DAYS_PER_MONTH = 26;
const TACTIVE_MONTHLY_COST_PER_CRORE = 15000; // ₹15,000 per crore of budget per month

/**
 * Calculate all losses based on project inputs
 * @param {object} input - { budget, duration_months, num_laborers }
 * @returns {object} Full calculation result
 */
export function calculateLosses(input) {
  const { budget, duration_months, num_laborers } = input;

  // ── Calculate Individual Losses ──
  const materialWaste = budget * LOSS_RATES.materialWaste;
  const idleMachinery = budget * LOSS_RATES.idleMachinery;

  // Labor inefficiency = laborers × daily wage × working days × duration × inefficiency rate
  const totalLaborCost = num_laborers * AVG_DAILY_WAGE * WORKING_DAYS_PER_MONTH * duration_months;
  const laborInefficiency = totalLaborCost * LOSS_RATES.laborInefficiency;

  const scheduleOverrun = budget * LOSS_RATES.scheduleOverrun;
  const reworkDefects = budget * LOSS_RATES.reworkDefects;

  const totalLoss = materialWaste + idleMachinery + laborInefficiency + scheduleOverrun + reworkDefects;

  // ── Calculate Tactive Recovery ──
  const materialRecovery = materialWaste * RECOVERY_RATES.materialTracking;
  const machineryRecovery = idleMachinery * RECOVERY_RATES.equipmentAlerts;
  const laborRecovery = laborInefficiency * RECOVERY_RATES.laborDashboard;
  const scheduleRecovery = scheduleOverrun * RECOVERY_RATES.scheduleOptimizer;
  const reworkRecovery = reworkDefects * RECOVERY_RATES.qualityControl;

  const totalRecovery = materialRecovery + machineryRecovery + laborRecovery + scheduleRecovery + reworkRecovery;

  // ── ROI Calculation ──
  const tactiveCost = (budget / 10000000) * TACTIVE_MONTHLY_COST_PER_CRORE * duration_months;
  const netSavings = totalRecovery - tactiveCost;
  const roiPercentage = tactiveCost > 0 ? ((netSavings / tactiveCost) * 100) : 0;
  const monthlySavings = totalRecovery / duration_months;
  const paybackMonths = monthlySavings > 0 ? Math.ceil(tactiveCost / monthlySavings) : 0;

  return {
    input: { budget, duration_months, num_laborers },

    losses: {
      materialWaste: {
        amount: materialWaste,
        percentage: LOSS_RATES.materialWaste * 100,
        label: 'Material Wastage',
        icon: '📦',
        description: 'Excess ordering, theft, damage, and poor inventory management',
      },
      idleMachinery: {
        amount: idleMachinery,
        percentage: LOSS_RATES.idleMachinery * 100,
        label: 'Idle Machinery',
        icon: '🏗️',
        description: 'Equipment sitting unused due to poor scheduling and maintenance',
      },
      laborInefficiency: {
        amount: laborInefficiency,
        percentage: ((laborInefficiency / budget) * 100),
        label: 'Labor Inefficiency',
        icon: '👷',
        description: 'Idle time, over-staffing, lack of productivity tracking',
      },
      scheduleOverrun: {
        amount: scheduleOverrun,
        percentage: LOSS_RATES.scheduleOverrun * 100,
        label: 'Schedule Overruns',
        icon: '⏰',
        description: 'Delays cascading across project phases and milestones',
      },
      reworkDefects: {
        amount: reworkDefects,
        percentage: LOSS_RATES.reworkDefects * 100,
        label: 'Rework & Defects',
        icon: '🔧',
        description: 'Quality issues requiring demolition and reconstruction',
      },
    },

    recovery: {
      materialTracking: {
        amount: materialRecovery,
        rate: RECOVERY_RATES.materialTracking * 100,
        module: 'Material Tracking',
        icon: '📦',
        description: 'Real-time inventory monitoring, automated reorder alerts, and wastage detection',
      },
      equipmentAlerts: {
        amount: machineryRecovery,
        rate: RECOVERY_RATES.equipmentAlerts * 100,
        module: 'Equipment Alert System',
        icon: '🏗️',
        description: 'Utilization tracking, predictive maintenance, and idle equipment reassignment',
      },
      laborDashboard: {
        amount: laborRecovery,
        rate: RECOVERY_RATES.laborDashboard * 100,
        module: 'Labor Productivity Dashboard',
        icon: '👷',
        description: 'Attendance tracking, zone-wise productivity scoring, and workforce optimization',
      },
      scheduleOptimizer: {
        amount: scheduleRecovery,
        rate: RECOVERY_RATES.scheduleOptimizer * 100,
        module: 'Schedule Optimizer',
        icon: '📅',
        description: 'AI-powered scheduling, critical path analysis, and delay prediction',
      },
      qualityControl: {
        amount: reworkRecovery,
        rate: RECOVERY_RATES.qualityControl * 100,
        module: 'Quality Control Monitor',
        icon: '✅',
        description: 'Inspection checklists, defect tracking, and compliance reporting',
      },
    },

    summary: {
      totalLoss,
      totalRecovery,
      tactiveCost,
      netSavings,
      roiPercentage,
      monthlySavings,
      paybackMonths,
      totalLaborCost,
    },
  };
}

/**
 * Get loss categories as array for chart rendering
 * @param {object} result - Output from calculateLosses
 * @returns {Array}
 */
export function getLossArray(result) {
  return Object.values(result.losses).map(loss => ({
    label: loss.label,
    amount: loss.amount,
    percentage: loss.percentage,
    icon: loss.icon,
  }));
}

/**
 * Get recovery modules as array
 * @param {object} result - Output from calculateLosses
 * @returns {Array}
 */
export function getRecoveryArray(result) {
  return Object.values(result.recovery).map(rec => ({
    module: rec.module,
    amount: rec.amount,
    rate: rec.rate,
    icon: rec.icon,
    description: rec.description,
  }));
}

File: backend/calculator.py
"""
Tactix Savings Calculation Engine.

Uses industry-standard loss percentages and Tactix module recovery rates
to estimate potential savings for construction projects.
"""

from models import (
    CalculationResult,
    LossBreakdown,
    ProjectInput,
    RecoveryReport,
)

# ---------------------------------------------------------------------------
# Industry-average loss percentages (as fraction of total budget)
# ---------------------------------------------------------------------------
MATERIAL_WASTE_PCT = 0.12        # 12% of budget lost to material wastage
IDLE_MACHINERY_PCT = 0.065       # 6.5% of budget lost to idle machinery
LABOR_INEFFICIENCY_DAILY = 800   # ₹800 per laborer per day lost to inefficiency
LABOR_INEFFICIENCY_PCT = 0.20    # 20% of labor-days are inefficient
SCHEDULE_OVERRUN_PCT = 0.15      # 15% of budget lost to schedule overruns
REWORK_DEFECTS_PCT = 0.07        # 7% of budget lost to rework & defects

# ---------------------------------------------------------------------------
# Tactix module recovery rates (fraction of respective loss recovered)
# ---------------------------------------------------------------------------
MATERIAL_TRACKING_RECOVERY = 0.70    # Material Tracking recovers 70%
EQUIPMENT_ALERTS_RECOVERY = 0.80     # Equipment Alerts recovers 80%
LABOR_DASHBOARD_RECOVERY = 0.60      # Labor Dashboard recovers 60%
SCHEDULE_OPTIMIZER_RECOVERY = 0.50   # Schedule Optimizer recovers 50%
QUALITY_CONTROL_RECOVERY = 0.65      # Quality Control recovers 65%

# ---------------------------------------------------------------------------
# Tactix subscription cost estimate (annual)
# Used to compute payback period.
# ---------------------------------------------------------------------------
TACTIX_ANNUAL_COST_RATIO = 0.015  # ~1.5% of project budget per year


def _round2(value: float) -> float:
    """Round a value to 2 decimal places."""
    return round(value, 2)


def calculate_losses(project: ProjectInput) -> LossBreakdown:
    """
    Calculate the estimated losses for a construction project
    based on industry averages.

    Args:
        project: The project parameters.

    Returns:
        A LossBreakdown with itemised and total losses.
    """
    budget = project.budget
    duration_days = project.duration_months * 30  # approximate days

    material_waste = budget * MATERIAL_WASTE_PCT
    idle_machinery = budget * IDLE_MACHINERY_PCT
    labor_inefficiency = (
        project.num_laborers
        * LABOR_INEFFICIENCY_DAILY
        * duration_days
        * LABOR_INEFFICIENCY_PCT
    )
    schedule_overrun = budget * SCHEDULE_OVERRUN_PCT
    rework_cost = budget * REWORK_DEFECTS_PCT

    total_loss = (
        material_waste
        + idle_machinery
        + labor_inefficiency
        + schedule_overrun
        + rework_cost
    )

    return LossBreakdown(
        material_waste=_round2(material_waste),
        idle_machinery=_round2(idle_machinery),
        labor_inefficiency=_round2(labor_inefficiency),
        schedule_overrun=_round2(schedule_overrun),
        rework_cost=_round2(rework_cost),
        total_loss=_round2(total_loss),
    )


def calculate_recovery(losses: LossBreakdown) -> RecoveryReport:
    """
    Calculate recoverable savings using Tactix modules.

    Args:
        losses: The loss breakdown computed for the project.

    Returns:
        A RecoveryReport with itemised recoveries and ROI.
    """
    material_recovery = losses.material_waste * MATERIAL_TRACKING_RECOVERY
    machinery_recovery = losses.idle_machinery * EQUIPMENT_ALERTS_RECOVERY
    labor_recovery = losses.labor_inefficiency * LABOR_DASHBOARD_RECOVERY
    schedule_recovery = losses.schedule_overrun * SCHEDULE_OPTIMIZER_RECOVERY
    rework_recovery = losses.rework_cost * QUALITY_CONTROL_RECOVERY

    total_recovery = (
        material_recovery
        + machinery_recovery
        + labor_recovery
        + schedule_recovery
        + rework_recovery
    )

    roi_percentage = (
        (total_recovery / losses.total_loss * 100) if losses.total_loss > 0 else 0.0
    )

    return RecoveryReport(
        material_recovery=_round2(material_recovery),
        machinery_recovery=_round2(machinery_recovery),
        labor_recovery=_round2(labor_recovery),
        schedule_recovery=_round2(schedule_recovery),
        rework_recovery=_round2(rework_recovery),
        total_recovery=_round2(total_recovery),
        roi_percentage=_round2(roi_percentage),
    )


def calculate_payback_months(
    project: ProjectInput, total_recovery: float
) -> float:
    """
    Estimate payback period for Tactix subscription.

    The annual subscription cost is approximated as a percentage of
    the project budget. Payback = (annual_cost / 12) months until
    cumulative recovery exceeds cost.

    Args:
        project: The project parameters.
        total_recovery: Total recovery over the project lifetime.

    Returns:
        Payback period in months (capped at project duration).
    """
    annual_cost = project.budget * TACTIX_ANNUAL_COST_RATIO
    monthly_cost = annual_cost / 12
    monthly_recovery = (
        total_recovery / project.duration_months
        if project.duration_months > 0
        else 0
    )

    if monthly_recovery <= monthly_cost:
        return float(project.duration_months)

    # months until cumulative net recovery ≥ 0
    payback = monthly_cost / (monthly_recovery - monthly_cost) if (monthly_recovery - monthly_cost) > 0 else float(project.duration_months)
    return _round2(min(payback, project.duration_months))


def run_calculation(project: ProjectInput) -> CalculationResult:
    """
    Run the full Tactix savings calculation pipeline.

    Args:
        project: The project input parameters.

    Returns:
        Complete CalculationResult with losses, recovery, savings and payback.
    """
    losses = calculate_losses(project)
    recovery = calculate_recovery(losses)
    net_savings = _round2(recovery.total_recovery)
    payback_months = calculate_payback_months(project, recovery.total_recovery)

    return CalculationResult(
        input_data=project,
        loss_breakdown=losses,
        recovery_report=recovery,
        net_savings=net_savings,
        payback_months=payback_months,
    )


# ---------------------------------------------------------------------------
# Indian currency formatting helper
# ---------------------------------------------------------------------------

def format_inr(amount: float) -> str:
    """
    Format an amount in Indian currency notation.

    - Amounts >= 1 Crore  → "₹X.XX Crores"
    - Amounts >= 1 Lakh   → "₹X.XX Lakhs"
    - Otherwise           → "₹X,XXX" with Indian grouping

    Args:
        amount: The amount in INR.

    Returns:
        Human-readable string in Indian notation.
    """
    abs_amount = abs(amount)
    sign = "-" if amount < 0 else ""

    if abs_amount >= 1_00_00_000:  # 1 Crore
        value = abs_amount / 1_00_00_000
        return f"{sign}₹{value:,.2f} Crores"
    elif abs_amount >= 1_00_000:  # 1 Lakh
        value = abs_amount / 1_00_000
        return f"{sign}₹{value:,.2f} Lakhs"
    else:
        # Indian grouping: last 3 digits, then groups of 2
        s = f"{abs_amount:,.2f}"
        return f"{sign}₹{s}"


def format_inr_pdf(amount: float) -> str:
    """
    Format an amount in Indian currency notation for PDF output.

    Uses 'Rs.' instead of '₹' so the text renders correctly with
    fpdf2's built-in Helvetica font (which only supports latin-1).

    Args:
        amount: The amount in INR.

    Returns:
        Human-readable string in Indian notation using 'Rs.'.
    """
    abs_amount = abs(amount)
    sign = "-" if amount < 0 else ""

    if abs_amount >= 1_00_00_000:  # 1 Crore
        value = abs_amount / 1_00_00_000
        return f"{sign}Rs.{value:,.2f} Crores"
    elif abs_amount >= 1_00_000:  # 1 Lakh
        value = abs_amount / 1_00_000
        return f"{sign}Rs.{value:,.2f} Lakhs"
    else:
        s = f"{abs_amount:,.2f}"
        return f"{sign}Rs.{s}"
